app/models/spotlight/resource.rb

```diff
@@ -100,6 +102,19 @@ def touch_exhibit!
       def write?
         Spotlight::Engine.config.writable_index
       end
+
+      def cleanup_solr_record
+        blacklight_solr.delete_by_id(document_ids, params: { softCommit: true })
+      end
+
+      def document_ids
+        document_builder.documents_to_index.to_a.map { |y| y[:id] }
```

Comment:
Maybe your second suggestion is the better approach as to not making to many changes.